torchvision/prototype/transforms/_transform.py

```diff
@@ -37,9 +38,30 @@ def forward(self, *inputs: Any) -> Any:
 
         params = self._get_params(flat_inputs)
 
-        flat_outputs = [
-            self._transform(inpt, params) if check_type(inpt, self._transformed_types) else inpt for inpt in flat_inputs
-        ]
+        # This is a heuristic on how to deal with simple tensor inputs:
+        # 1. If we haven't seen an image yet, we transform a simple tensor
+        # 2. If we have seen an image before, so either a simple tensor, a `datapoints.Image` or a `PIL.Image.Image`,
+        #    we return simple tensors without modification.
+        # The order is defined by the returned list of `tree_flatten`, which recurses depth-first through the input.
+        # Since in most cases the image is the first input or at least comes before any other numerical data, this
+        # heuristic allows users to keep any supplemental numerical data in the sample as simple tensors. We have a few
+        # datasets, like `Caltech101`, `CelebA`, and `Widerface`, that would need special handling without this
+        # heuristic, since they return the target partially or completely as tensors.
```

Comment:
I think we should also explain that it comes from the constraint that a) we need to keep BC so single tensors must be treated as image and b) we still don't want to treat all tensors as image because that's too strong of an assumption for some datasets which output tensors that aren't images.

> Since in most cases the image is the first input or at least comes before any other numerical data, 

I'd suggest to put this part in a separate sentence at the end (after the explanation of why we need this), and continue it with

> Since in most cases the image is the first input or at least comes before any other numerical data, this heuristic should work well in practice